blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+
+Here's what was going on there:
+
+When the user resizes the window, Chromium needs to redraw the contents of the window for the new window size. This takes some time. The new frame isn't ready immediately.
+
+Here's a sequence of frames that demonstrates this:
+
+![First frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-1.png)
+![Second frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-2.png)
+![Third frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-3.png)
+![Fourth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-4.png)
+![Fifth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-5.png)
+
+At a certain time during the resize, Windows tells us: "The window is 1,000 pixels wide." But the frames that the browser compositor produces are lagging behind. The last frame that we have painted might be 600 pixels wide.
+
+Historically, Chromium used to skip frames where the width of the window did not match the width of the frame that it last painted. It would decide to just not update the window.
+
+However, that would often lead to the window contents not being updated at all until the resize operation was finished.
+
+So [in 2015](https://crrev.com/1426453006) someone decided: "Why not show these frames? They might not match the window size perfectly, but at least we can show _something_."
+
+It would lead to gutter, but at that time the gutter was black. So that was better than the previous implementation.
+
+Now, 10 years later with DirectComposition, that gutter was often filled with stale pixels.
+
+Let's look at what was happening there:
+
+Every frame consists of multiple render passes. The render passes represent the various things that should be drawn on screen. From complicated bitmaps to rectangles filled with solid colors.
+
+Every frame has a root render pass, which contains all other render passes and glues them together. (Render passes are arranged in a tree structure and the root render pass is the root of that tree.)
+
+So now during a resize, we'd get to a point where we know the window is 1,000 pixels wide. Accordingly, we'd adjust the viewport of our output surface to also be 1,000 pixels wide. But the frame that we just received is only 600 pixels wide.
+
+The [optimization from 2015](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/display.cc;l=1053-1062;drc=152a83bb9fcba4ea3a7b9f8efbe0f6204f95eadd) would then go and change the width of the root render pass to also be 1,000 pixels. But it wouldn't change what the render passes would actually draw on screen. They'd still only contain instructions to draw a picture that is 600 pixels wide.
```

Comment:
```suggestion
The [optimization from 2015](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/display.cc;l=1053-1062;drc=152a83bb9fcba4ea3a7b9f8efbe0f6204f95eadd) would then go and change the width of the root render pass to also be 1,000 pixels, but wouldn't change what the render passes would actually draw on screen. They'd still only contain instructions to draw a picture that is 600 pixels wide.
```

Reply:
School teachers tell us not to start sentences with "but."

But there's [no actual grammar rule against it](https://prowritingaid.com/starting-a-sentence-with-and-or-but). :)

Sometimes I like it to avoid long sentences. Short sentences are much easier to read.